File: apps/web/src/services/chat-stream.ts
import type { Compact } from '../api.js';

export type RunEvent = { type: 'started' | 'delta' | 'usage' | 'completed' | 'failed' | 'cancelled'; runId: string; seq: number; text?: string; error?: string; result?: Compact; contextTruncated?: boolean };

export async function chatStream(branchId: string, revision: number, signal: AbortSignal, receive: (event: RunEvent) => void): Promise<Compact> {
  const response = await fetch('/api/ai/chat/stream', {
    method: 'POST', credentials: 'same-origin', signal,
    headers: { 'content-type': 'application/json' }, body: JSON.stringify({ branchId, revision, response: 'compact' }),
  });
  if (!response.ok) throw new Error((await response.json()).error ?? '请求失败。');
  if (!response.body) throw new Error('浏览器不支持流式响应。');
  const reader = response.body.getReader(), decoder = new TextDecoder('utf-8', { fatal: true });
  let buffer = '', runId = '', seq = 0;
  try {
    while (true) {
      const { value, done } = await reader.read();
      buffer += decoder.decode(value, { stream: !done });
      let end: number;
      while ((end = buffer.indexOf('\n\n')) >= 0) {
        const raw = buffer.slice(0, end); buffer = buffer.slice(end + 2);
        if (!raw.startsWith('data: ')) continue;
        const event = JSON.parse(raw.slice(6)) as RunEvent;
         if ((!runId && event.type !== 'started') || (runId && event.runId !== runId) || !Number.isSafeInteger(event.seq) || event.seq <= seq || event.seq > 1000000) throw new Error('模型流顺序无效。');
        runId = event.runId; seq = event.seq; receive(event);
        if (event.type === 'failed') throw new Error(event.error ?? '模型请求失败。');
        if (event.type === 'cancelled') throw new DOMException('已停止生成', 'AbortError');
         if (event.type === 'completed' && event.result) return event.result;
      }
       if (buffer.length > 1024 * 1024) throw new Error('响应事件超过限制。');
      if (done) throw new Error('模型流意外中断，问题已保留。');
    }
  } finally { await reader.cancel().catch(() => {}); reader.releaseLock(); }
}

File: backend/app/chat_stream.py
"""Request-owned runs: no model I/O inside database transactions."""

import asyncio
import hashlib
import json
import uuid

from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session

from .db import engine
from .providers import events
from .services import chat_request, context_messages, error

# This application supports one backend process, including desktop.
_runs: dict[tuple[str, str], str] = {}
_tasks = {}


def signature(branch):
    # UI-only edits do not invalidate generation, while context edits do.
    content = {
        k: branch.get(k) for k in ("entries", "awaiting", "selection", "pendingPrompt")
    }
    return hashlib.sha256(
        json.dumps(content, sort_keys=True, ensure_ascii=False).encode()
    ).hexdigest()


def start(input, uid, database, snapshot, replace, transition):
    from .repository import Repository
    repo = Repository(database, uid)
    current = repo.view()
    if input.revision != current["revision"]:
        error(409, "工作区已更新，请刷新后重试。")
    branch = repo.context(input.branchId)
    if (
        not branch
        or not branch.get("awaiting")
        or not branch["entries"]
        or branch["entries"][-1]["role"] != "user"
    ):
        error(409, "没有等待回答的问题。")
    key = (uid, input.branchId)
    if key in _runs:
        error(409, "此主题已在生成回答。")
    request = chat_request(
        database,
        uid,
        context_messages(branch),
    )
    expected = signature(branch)
    database.rollback()
    run_id = str(uuid.uuid4())
    _runs[key] = run_id

    async def body():
        parts, last_seq = [], 0
        _tasks[run_id] = asyncio.current_task()
        stream = events(request, run_id=run_id)
        try:
            async for item in stream:
                last_seq = item["seq"]
                if item["type"] == "started":
                    item["contextTruncated"] = branch["contextTruncated"]
                if _runs.get(key) != run_id:
                    yield encode(
                        {"type": "cancelled", "runId": run_id, "seq": last_seq}
                    )
                    return
                if item["type"] == "delta":
                    parts.append(item["text"])
                if item["type"] == "completed":
                    answer = "".join(parts)
                    if not answer.strip():
                        error(502, "模型返回空回答。")
                    with Session(engine) as writer:
                        repository = Repository(writer, uid)
                        latest = repository.view()
                        target = repository.context(input.branchId)
                        if target is None or signature(target) != expected:
                            error(409, "主题上下文已更新，此次回答未保存。")
                        result = repository.action(
                            {
                                "type": "answer",
                                "branchId": input.branchId,
                                "text": answer,
                                "revision": latest["revision"],
                            },
                            compact=getattr(input, "response", "compact") == "compact",
                        )
                        item["result" if getattr(input, "response", "compact") == "compact" else "snapshot"] = result
                yield encode(item)
        except asyncio.CancelledError:
            raise
        except Exception as exc:
            yield encode(
                {
                    "type": "failed",
                    "runId": run_id,
                    "seq": last_seq + 1,
                    "error": getattr(exc, "message", "回答保存失败，请刷新后重试。"),
                }
            )
        finally:
            await stream.aclose()
            _tasks.pop(run_id, None)
            if _runs.get(key) == run_id:
                _runs.pop(key, None)

    return StreamingResponse(
        body(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )


def cancel(uid, branch_id, run_id):
    key = (uid, branch_id)
    if _runs.get(key) == run_id:
        _runs.pop(key, None)
        task = _tasks.get(run_id)
        if task and task is not asyncio.current_task():
            task.cancel()
    return {"ok": True}


def encode(item):
    return "data: " + json.dumps(item, ensure_ascii=False) + "\n\n"
